### PR TITLE
Avoid __dict__ and __weakref__ attributes in AnyUrl and _BaseAddress subclasses

### PR DESCRIPTION
**Pull request by nuno-andre**
## Change Summary

Add an empty `__slots__` declaration to `AnyUrl` and `_BaseAddress` subclasses to avoid the creation of `__dict__` and `__weakreef__` attributes. When inheriting from a class with `__slots__`, [child subclasses will get a `__dict__` and `__weakref__` unless they also define `__slots__`](https://docs.python.org/3/reference/datamodel.html#notes-on-using-slots).

```python
from pydantic import AnyUrl, AnyHttpUrl

class AnyOtherHttpUrl(AnyUrl):
    allowed_schemes = {'http', 'https'}
    __slots__ = ()

print(set(dir(AnyHttpUrl)) - set(dir(AnyUrl)))
#> {'__weakref__', '__dict__'}

print(set(dir(AnyOtherHttpUrl)) - set(dir(AnyUrl)))
#> set()
```

## Related issue number

None.

## Checklist

* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI and coverage remains at 100%
* [ ] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by PrettyWood** [comment]
Hi @nuno-andre 
What's the matter with having `__dict__`  and `__weakref__`?
What problem does it solve?
AFAIR I haven't seen any issue related to this 🤔

**Comment by nuno-andre** [comment]
Hi @PrettyWood!

One of the `__slots__`'s purposes is [to deny the creation of `__dict__` to save memory](https://docs.python.org/3/reference/datamodel.html#slots).

```python
from pydantic import AnyUrl, AnyHttpUrl
from pympler.asizeof import asizeof


class AnyOtherHttpUrl(AnyUrl):
    allowed_schemes = {'http', 'https'}
    __slots__ = ()


a = AnyUrl.__new__(AnyHttpUrl, 'https://example.com')
print(asizeof(a))
#> 224

b = AnyUrl.__new__(AnyOtherHttpUrl, 'https://example.com')
print(asizeof(b))
#> 120
```

```python
from pydantic import BaseModel


class A(BaseModel):
    url: AnyHttpUrl


class B(BaseModel):
    url: AnyOtherHttpUrl


a = A(url='https://example.com')
print(asizeof(a))
#> 808

b = B(url='https://example.com')
print(asizeof(b))
#> 704
```

I think this is a pretty straightforward and valuable optimization and also makes a more consistent type model.

File: changes/2890-nuno-andre.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Avoid __dict__ and __weakref__ attributes in AnyUrl and _BaseAddress subclasses.
```
Comment (samuelcolvin) [inline changes/2890-nuno-andre.md:1]:
```suggestion
Avoid `__dict__` and `__weakref__` attributes in `AnyUrl` and IP address fields.
```

**Review by samuelcolvin — COMMENTED** [review]
Seems good to me, I can't see how this will break anyone's code. But I'm sure we'll have someone shouting at us about breaking code in a few weeks from some weird usage.

**Comment by tl24** [comment]
Let the shouting begin... 😃 We had a breaking change as we were reading the slots to be able to clone a RedisDsn and add in a password that was stored in a secret.  We worked around by hard-coding the property names, but is there a better way of doing that?

For reference:
```
def copy_dsn(original, **override_args):
    # object properties don't show up with normal methods,
    # so we need to use the __slots__ property to get field names
    attrs = {n: getattr(original, n) for n in original.__slots__ if getattr(original, n) is not None}
    return type(original)(None, **{**attrs, **override_args})

dsn_with_password = copy_dsn(RedisDsn("redis://localhost:6379/0", password="secret")
```

**Comment by tl24** [comment]
I don't know if this would be a frequent use case, but something like copy on BaseModel would be nice:
```
dsn_with_password = RedisDsn("redis://localhost:6379/0").copy(update={"password":"secret"})
```

**Pull request merged**